src/Symfony/Bundle/FrameworkBundle/Client.php

```diff
@@ -104,6 +106,45 @@ public function enableReboot()
         $this->reboot = true;
     }
 
+    /**
+     * Logins a user against a specific firewall.
+     *
+     * @param string|object $user
+     * @param string        $credentials
+     * @param string[]      $roles
+     * @param string        $firewallName
+     * @param string|null   $firewallContext
+     */
+    public function login(
+        $user,
+        string $credentials,
+        array $roles = array('ROLE_ADMIN'),
+        string $firewallName = 'main',
+        string $firewallContext = null
+    ) {
+        if (!class_exists(UsernamePasswordToken::class)) {
+            throw new \RuntimeException('You must install the Symfony Security component to use this feature.');
+        }
+
+        $context = $firewallContext ?? $firewallName;
+        $session = $this->getContainer()->get('session');
+
+        $token = new UsernamePasswordToken($user, $credentials, $firewallName, $roles);
+        $session->set('_security_'.$context, serialize($token));
+        $session->save();
+
+        $cookie = new Cookie($session->getName(), $session->getId());
+        $this->getCookieJar()->set($cookie);
+    }
+
+    /**
+     * Logouts a user.
+     */
+    public function logout()
+    {
+        $this->getContainer()->get('session')->invalidate();
```

Comment:
We don't need to invalidate the whole session. IMO we need to use the configured firewall's behavior when logging out, which can decide whether to only log out without invalidating the session.

Reply:
@Pierstoval we cannot call the firewall logic, as we are not in the context of handling a HttpFoundation request (so for instance, invalidating a remember me cookie would not work, unless we do complex logic to emulate things, and then we might as well call the logout endpoint rather than duplicating all the code)

Reply:
Fair point, defaulting to session invalidation might be better, the other one possibly being a simple HTTP call to the logout endpoint